tools/daemon/nl.c

```diff
@@ -0,0 +1,239 @@
+
+void nl_destroy(nl_t nt)
+{
+	if (nt) {
+		close(nt->fd);
+		free(nt);
+		nt = NULL;
+	}
+}
+
+int nl_send(nl_t nt, struct nlmsghdr *nl_msg)
+{
+	struct sockaddr_nl nladdr;
+	struct iovec iov;
+	struct msghdr msg;
+	int ret = -1;
+
+	nl_msg->nlmsg_seq = nt->seq++;
+
+	memset(&nladdr, 0, sizeof(nladdr));
+	nladdr.nl_family = AF_NETLINK;
+	nladdr.nl_pid = 0;
+	nladdr.nl_groups = 0;
+
+	iov.iov_base = nl_msg;
+	iov.iov_len = nl_msg->nlmsg_len;
+
+	memset(&msg, 0, sizeof(msg));
+	msg.msg_name = &nladdr;
+	msg.msg_namelen = sizeof(nladdr);
+	msg.msg_iov = &iov;
+	msg.msg_iovlen = 1;
+
+	log_hexdump((void *)nl_msg, nl_msg->nlmsg_len);
+	ret = sendmsg(nt->fd, &msg, 0);
+	if (ret < 0) {
+		log_error("Failed to send netlink message: %s (%d)\n",
+			strerror(errno), errno);
+		return ret;
+	}
+
+	return ret;
+}
+
+int nl_recv(nl_t nt, int (*cb)(struct nlmsghdr *, void *arg), void *arg)
+{
+	struct sockaddr_nl nladdr;
+	struct iovec iov;
+	struct msghdr msg;
+	int ret = 0;
+	int multipart = 0;
+
+	memset(&nladdr, 0, sizeof(nladdr));
+
+	iov.iov_base = nt->buf;
+	iov.iov_len = sizeof(nt->buf);
+
+	memset(&msg, 0, sizeof(msg));
+	msg.msg_name = &nladdr;
+	msg.msg_namelen = sizeof(nladdr);
+	msg.msg_iov = &iov;
+	msg.msg_iovlen = 1;
+
+	do {
+		struct nlmsghdr *nl_msg;
+		int recv_bytes = 0;
+
+		recv_bytes = recvmsg(nt->fd, &msg, 0);
+		if (recv_bytes <= 0) {
+			if (errno == EAGAIN || errno == EWOULDBLOCK || errno == EINTR) {
+				continue;
+			}
+			return -1;
+		}
+
+		for (nl_msg = (struct nlmsghdr *)nt->buf;
+		     NLMSG_OK(nl_msg, (unsigned int)recv_bytes);
+		     nl_msg = NLMSG_NEXT(nl_msg, recv_bytes)) {
+			if (nl_msg->nlmsg_type == NLMSG_ERROR) {
+				struct nlmsgerr *err_data = NLMSG_DATA(nl_msg);
+
+				if (err_data->error < 0) {
+					errno = -err_data->error;
+					return -1;
+				}
+				/* Ack message. */
+				return 0;
+			}
+			/* Multi-part msgs and their trailing DONE message. */
+			if (nl_msg->nlmsg_flags & NLM_F_MULTI) {
+				if (nl_msg->nlmsg_type == NLMSG_DONE) {
+					return 0;
+				}
+				multipart = 1;
+			}
+			if (cb) {
+				ret = cb(nl_msg, arg);
+			}
+		}
+	} while (multipart || (msg.msg_flags & MSG_TRUNC));
+
+	return ret;
+}
+
+void nl_attr_add(struct nlmsghdr *nl_msg, unsigned short type,
+		const void *data, unsigned int data_len)
+{
+	struct rtattr *rta;
+
+	if ((NLMSG_ALIGN(nl_msg->nlmsg_len) + RTA_ALIGN(RTA_LENGTH(data_len))) > NLMSG_BUF) {
+		log_error("Message size is: %d that exceeds limit: %d\n",
+				(NLMSG_ALIGN(nl_msg->nlmsg_len) + RTA_ALIGN(RTA_LENGTH(data_len))), NLMSG_BUF);
+		return ;
+	}
+	rta = (struct rtattr *)NLMSG_TAIL(nl_msg);
+	rta->rta_len = RTA_LENGTH(data_len);
+	rta->rta_type = type;
+	if (data && data_len) {
+		memcpy(RTA_DATA(rta), data, data_len);
+	}
+	nl_msg->nlmsg_len = NLMSG_ALIGN(nl_msg->nlmsg_len) + RTA_ALIGN(rta->rta_len);
+}
+
+struct rtattr *nl_attr_nest_start(struct nlmsghdr *nl_msg, int type)
+{
+	struct rtattr *nest = NLMSG_TAIL(nl_msg);
+
+	nl_attr_add(nl_msg, type, NULL, 0);
+
```

Comment:
need to verify that no need to malloc new element here (and free in nl_attr_nest_end)

Reply:
no need because all are done on stack

Reply:
it is done on stack in current implementation